saleor/graphql/webhook/subscription_query.py

```diff
@@ -0,0 +1,146 @@
+from graphql.error import GraphQLSyntaxError
+from graphql.language.ast import (
+    Document,
+    Field,
+    FragmentDefinition,
+    FragmentSpread,
+    InlineFragment,
+    OperationDefinition,
+)
+
+from saleor.webhook.error_codes import WebhookErrorCode
+
+
+class IsFragment(Flag):
+    TRUE = True
+    FALSE = False
+
+
+class SubscriptionQuery:
+    def __init__(self, query: str):
+        self.query: str = query
+        self.is_valid: bool = False
+        self.ast: Document = Document("")
+        self.events: List[str] = []
+        self.error_code: Optional[str] = None
+        self.errors = self.validate_query()
+        self.error_msg: str = ";".join(set([err.message for err in self.errors]))
+
+    def validate_query(self):
+        from ..api import schema
+
+        graphql_backend = get_default_backend()
+        try:
+            document = graphql_backend.document_from_string(schema, self.query)
+            self.ast = document.document_ast
+            errors = validate(schema, self.ast)
+        except GraphQLSyntaxError as e:
+            self.error_code = WebhookErrorCode.SYNTAX.value
+            return [e]
+
+        if errors:
+            self.error_code = WebhookErrorCode.GRAPHQL_ERROR.value
+            return errors
+
+        try:
+            self.events = self.get_events_from_subscription()
+        except ValidationError as err:
+            self.error_code = err.code
+            return [err]
+
+        self.is_valid = True
+        return []
+
+    def get_events_from_subscription(self) -> List[str]:
+        subscription = self._get_subscription(self.ast)
+        if not subscription:
+            raise ValidationError(
+                message="Subscription operation can't be found.",
+                code=WebhookErrorCode.MISSING_SUBSCRIPTION.value,
+            )
+
+        event_types = self._get_event_types_from_subscription(subscription)
+        if not event_types:
+            raise ValidationError(
+                message="Event field can't be found.",
+                code=WebhookErrorCode.UNABLE_TO_PARSE.value,
+            )
+
+        events_and_fragments: Dict[str, IsFragment] = {}
+        for event_type in event_types:
+            self._get_events_from_field(event_type, events_and_fragments)
+
+        fragment_definitions = self._get_fragment_definitions(self.ast)
+        unpacked_events: Dict[str, IsFragment] = {}
+        for event_name, is_fragment in events_and_fragments.items():
+            if not is_fragment:
+                continue
+            event_definition = fragment_definitions[event_name]
+            self._get_events_from_field(event_definition, unpacked_events)
+        events_and_fragments.update(unpacked_events)
+        events = [k for k, v in events_and_fragments.items() if not v]
+
+        if not events:
+            raise ValidationError(
+                message="Can't find a single event.",
+                code=WebhookErrorCode.UNABLE_TO_PARSE.value,
```

Comment:
Wouldn't be good to have a separate error code here? We were able to determine that we don't have any events here, so we were able to parse the query, right?

Reply:
MISSING_EVENT error added